core/src/main/java/org/apache/iceberg/avro/ProjectionDatumReader.java

```diff
@@ -26,29 +26,33 @@
 import org.apache.avro.Schema;
 import org.apache.avro.io.DatumReader;
 import org.apache.avro.io.Decoder;
+import org.apache.iceberg.mapping.NameMapping;
 import org.apache.iceberg.types.TypeUtil;
 
 public class ProjectionDatumReader<D> implements DatumReader<D> {
   private final Function<Schema, DatumReader<?>> getReader;
   private final org.apache.iceberg.Schema expectedSchema;
   private final Map<String, String> renames;
+  private final NameMapping nameMapping;
   private Schema readSchema = null;
   private Schema fileSchema = null;
   private DatumReader<D> wrapped = null;
 
   public ProjectionDatumReader(Function<Schema, DatumReader<?>> getReader,
                                org.apache.iceberg.Schema expectedSchema,
-                               Map<String, String> renames) {
+                               Map<String, String> renames,
+                               NameMapping nameMapping) {
     this.getReader = getReader;
     this.expectedSchema = expectedSchema;
     this.renames = renames;
+    this.nameMapping = nameMapping;
   }
 
   @Override
   public void setSchema(Schema newFileSchema) {
     this.fileSchema = newFileSchema;
     Set<Integer> projectedIds = TypeUtil.getProjectedIds(expectedSchema);
-    Schema prunedSchema = AvroSchemaUtil.pruneColumns(newFileSchema, projectedIds);
+    Schema prunedSchema = AvroSchemaUtil.pruneColumns(newFileSchema, projectedIds, nameMapping);
```

Comment:
@rdblue do you think it would make sense to initialize `nameMapping` with `expectedSchema` if it is null. This could help us read old Avro data where table does not have an external schema mapping specified.

Reply:
@rdblue any thoughts on this?

Reply:
I think I mentioned it elsewhere, but I think the behavior should be:

1. If there is a name mapping in table metadata, pass it in
2. Otherwise, if the incoming schema has no field IDs, infer a mapping from the current schema and use it
3. Otherwise the schema has at least one column ID, so no inferred mapping should be used

Reply:
Thanks. Is it ok if I address this as part of another commit?

Reply:
Yes, of course. I don't think that needs to be part of this one. This is just adding support for a mapping if you specify one.